### ISSUE TITLE
SEC-44: Liberty Reader role support

### ISSUE THREAD

**Issue by jvanhill** [OpenLiberty/open-liberty#7253]
https://bigblue.aha.io/features/SEC-44

Liberty today only offers a single administration role, 'Administrator'. There is no other option. Many customers have asked for this. This feature adds support for a new 'Viewer', or reader role.

See the WebSphere Liberty epic here: 

> https://github.ibm.com/was-liberty/WS-CD-Open/issues/15384

See the design documents here:
> [WAD Document](https://ibm.box.com/s/u0i8aiel8rtqxvjuerqmb34gppvwety0)
> [RESTHandler APIs](https://ibm.box.com/s/2luthyqr2p872truodwvrazxu94cg4ru)

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same BOX folder.  Thanks!

**Comment by marikaj123** [comment]
@jvanhill - Does it require translation?

**Comment by skasund** [comment]
Please reach out to @andrea-pichler for any STE approvals between 10/19-10/22.  I'll be back on 10/23.  Thanks!

**Comment by jvanhill** [comment]
> Please reach out to @andrea-pichler for any STE approvals between 10/19-10/22. I'll be back on 10/23. Thanks!

@skasund , @andrea-pichler - This feature is now targetted for 19.0.0.12.

**Comment by jvanhill** [comment]
> @jvanhill - Does it require translation?

@marikaj123 - I believe we had everything in prior to the translation deadline. I don't recall needing to add any translation messages, b/c REST responses use HTTP codes so there wasn't much need to add any WARNING or SEVERE messages that would need to be translated.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by jvanhill** [comment]
@donbourne -

(NOT READY FOR REVIEW, SAVING RESPONSE)

1) The main area of concern was the customer determining whether their user was in the reader-role. We have added auditing to the RESTHandlerContainerImpl that should make clear what role the requested operation required. This includes framework for all RESTHandler implementations to pass back what roles were required for the requested operation.

2) TBD

3a) Nicole Stewart
3b) TBD

4) WAS L2: SEC, WAS L3: Core Security

**Comment by gscottj** [comment]
For accessibility compliance, information about disable widgets needs to be provided to the user when assigned to the Reader role so they will be informed why they fields are disabled. The designs for this feature appear to be sufficient to provide that information using a combination of hover help on the input fields, error message stating the user role assignment is read only, and other indicators that show the person's role assignment "Reader" and that it is for Read Only access.

**Comment by chirp1** [comment]
Jesse created a doc issue in OL and another in WL. @jvanhill The OL docs for this feature will be prioritized on the backlog. The docs might not be written by the time that the function is available in 19.0.0.12. So, you can add enough info in your blog for the user to have an example of how ot use it.

**Comment by jvanhill** [comment]
@skasund -

I have uploaded the STE slides to the STE Archive folder in Box.
> STE_SEC-44_ReaderRole.ppt

**Comment by skasund** [comment]
@jvanhill  Thanks for the STE slides.  I've approved the feature.

**Comment by jvanhill** [comment]
This feature has made it into the 190012 GM. Closing.

**Comment by marikaj123** [comment]
Jesse, Thank you for confirming the feature did make the GM release 190012.
